specification/transport_layer/transport_layer.tex

```diff
@@ -575,6 +575,72 @@ \subsection{Transfer-ID comparison}\label{sec:transfer_id_forward_distance}
 \end{minted}
 \end{minipage}
 
+\subsection{Payload truncation}
+
+The deterministic nature of UAVCAN in general and DSDL in particular allows implementations to statically determine the
+maximum amount of memory that is required to contain a data object of a particular type.
+Consequently, an implementation that is interested in receiving data objects of a particular type\footnote{%
+    Messages, service requests, or service responses.
+}
+can statically determine the maximum length of the transfer payload.
+
+Implementations shall be able to handle incoming transfers containing a larger amount of payload data than expected.
```

Comment:
how does a receiver determine the "true" payload size? Is this always determined via the MTU? Does this requirement limit portability to other network layers?

Reply:
The true payload size is determined by deserializing the DSDL object. We no longer have the tail array optimization in v1, so the (de)serialization process is completely decoupled from the transport layer. This does not affect portability in any way.

Reply:
If I'm deserializing a type then the type's total length is all I know at the DSDL processor level. I can accept additional frames in a multi-frame transfer using the tail-byte but to find additional bits within a frame I need some other length field. In CAN this is the MTU/dlc. Is there some other way you were thinking the "true" size of a transfer can be determined? Given this we are requiring an MTU to be available at the transport layer. What we cannot determine, however, is what is additional payload and what is empty overhead of the network frame (e.g. CAN-FD padding) but I think this is okay if we always include all padding in the CRC (as we do now). My comment here is, in your research into other technologies have you come across an interesting network that allows bits to be added to a frame (i.e. within the length specified by the MTU) after a sender has enqueued a message?

Additionally, there is a physical limit to to this rule that should be acknowledged as being unspecified. i.e. there is some maximum RX buffer length beyond which the system will reject the transfer. For CAN-FD this would be a system originally designed to accept only (say) 24 bytes attached to a network sending semantically compatible 64 byte messages (The CRC can be accumulated byte-wise so we aren't limited by the total number of frames in a transfer). The sender would have no way of knowing the receivers limitation. This means we require all nodes to support the same MTU in order to enable payload truncation. Note that this is a real possibility given memory-constrained MCUs where allocating 64-byte RX buffers is prohibitive and, for example, where the same limitation doesn't constrain the larger processor in the flight controller.

Reply:
> Is there some other way you were thinking the "true" size of a transfer can be determined?

No. Here, the size only becomes known after deserialization, and it depends on the DSDL definition in use. As you wrote, the remaining unused data may be padding, sub-type fields, or both -- the receiver is unable to distinguish them.

> My comment here is, in your research into other technologies have you come across an interesting network that allows bits to be added to a frame (i.e. within the length specified by the MTU) after a sender has enqueued a message?

Not exactly, no. The closest behavior I am aware of is the min-size padding (like in Ethernet to guarantee a minimum packet propagation time for collision detection over long links) and cryptographic padding.

> Additionally, there is a physical limit to to this rule that should be acknowledged as being unspecified. i.e. there is some maximum RX buffer length beyond which the system will reject the transfer. For CAN-FD this would be a system originally designed to accept only (say) 24 bytes attached to a network sending semantically compatible 64 byte messages (The CRC can be accumulated byte-wise so we aren't limited by the total number of frames in a transfer). The sender would have no way of knowing the receivers limitation. This means we require all nodes to support the same MTU in order to enable payload truncation.

Yes, we will be forcing a trade-off upon system designers between extensibility and resource utilization. Computer networks typically expect the MTU to be the same across the link so the requirement is not new. Generally, you could perhaps see how an application that attempts to optimize RAM usage by limiting the MTU for a particular subject/service has a leaky abstraction spanning between the low-level network driver and the high-level data type definition.

Reply:
I just pushed a commit that I expect addresses this issue: https://github.com/UAVCAN/specification/pull/64/commits/b3893e367dbf9abe3058c1e07762f26dc20ee20d

> Additionally, there is a physical limit to to this rule that should be acknowledged as being unspecified

I did not acknowledge this explicitly. Instead, I worded this as a weak requirement ("should") with an elaboration about the motivation.